# Phase 2 — FD001 Clustering and Visualization

This block applies **KMeans** and **Agglomerative Clustering** to the cleaned `FD001` dataset using sensor features. It then performs:

- **Silhouette Score Evaluation** for clustering quality
- **Dimensionality Reduction** using **PCA** and **t-SNE**
- **Cluster Visualization** using scatterplots
- **Stage Distribution Plots** to analyze cluster spread
- **CSV Export** of the clustered data for downstream tasks

All visual outputs are saved to `figures/`, and processed data is stored as `clustered_train_FD001.csv`.


In [1]:
# Phase 2: FD001 Clustering with KMeans and Agglomerative
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score
import os

ds_id = 'FD001'
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

df = pd.read_csv(os.path.join(base_path, f"clean_train_{ds_id}.csv"))
sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
X = df[sensor_cols]

# Clustering
df['kmeans_cluster'] = KMeans(n_clusters=5, random_state=42).fit_predict(X)
df['agglo_cluster'] = AgglomerativeClustering(n_clusters=5).fit_predict(X)
df['kmeans_stage'] = df['kmeans_cluster'].apply(lambda n: f"Stage {n}")
df['agglo_stage'] = df['agglo_cluster'].apply(lambda n: f"Stage {n}")

# Silhouette
print(f"{ds_id} — KMeans Silhouette:", silhouette_score(X, df['kmeans_cluster']))
print(f"{ds_id} — Agglomerative Silhouette:", silhouette_score(X, df['agglo_cluster']))

# PCA
pca = PCA(n_components=2).fit_transform(X)
df['pca_1'], df['pca_2'] = pca[:, 0], pca[:, 1]
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='kmeans_stage')
plt.title(f"{ds_id}: KMeans via PCA")
plt.savefig(os.path.join(figures_path, f"{ds_id}_PCA_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='agglo_stage')
plt.title(f"{ds_id}: Agglomerative via PCA")
plt.savefig(os.path.join(figures_path, f"{ds_id}_PCA_Agglo.png"))
plt.close()

# t-SNE
tsne = TSNE(n_components=2, perplexity=30, random_state=42, n_iter=1000)
tsne_result = tsne.fit_transform(X)
df['tsne_1'], df['tsne_2'] = tsne_result[:, 0], tsne_result[:, 1]
sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='kmeans_stage')
plt.title(f"{ds_id}: KMeans via t-SNE")
plt.savefig(os.path.join(figures_path, f"{ds_id}_TSNE_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='agglo_stage')
plt.title(f"{ds_id}: Agglomerative via t-SNE")
plt.savefig(os.path.join(figures_path, f"{ds_id}_TSNE_Agglo.png"))
plt.close()

# Distribution
plt.figure(figsize=(10, 4))
sns.countplot(x='kmeans_stage', data=df, palette='tab10')
plt.title(f"{ds_id} KMeans Cluster Distribution")
plt.savefig(os.path.join(figures_path, f"{ds_id}_KMeans_Distribution.png"))
plt.close()

sns.countplot(x='agglo_stage', data=df, palette='tab10')
plt.title(f"{ds_id} Agglomerative Cluster Distribution")
plt.savefig(os.path.join(figures_path, f"{ds_id}_Agglo_Distribution.png"))
plt.close()

# Save
df.to_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"), index=False)
print(f"[✔] Saved clustered dataset: clustered_train_{ds_id}.csv")


FD001 — KMeans Silhouette: 0.18804918038535223
FD001 — Agglomerative Silhouette: 0.1759524903893289


c:\Users\pandr\hybrid-predictive-maintenance\venv\Lib\site-packages\sklearn\manifold\_t_sne.py:1164: FutureWarning: 'n_iter' was renamed to 'max_iter' in version 1.5 and will be removed in 1.7.
  warnings.warn(
C:\Users\pandr\AppData\Local\Temp\ipykernel_8512\560400801.py:61: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(x='kmeans_stage', data=df, palette='tab10')
C:\Users\pandr\AppData\Local\Temp\ipykernel_8512\560400801.py:66: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(x='agglo_stage', data=df, palette='tab10')


[✔] Saved clustered dataset: clustered_train_FD001.csv


#  Phase 2: Clustering and Visualization for FD002

This script performs **unsupervised clustering** of the `clean_train_FD002.csv` dataset using:

-  **KMeans** and **Agglomerative Clustering** (5 clusters → Stage 0 to Stage 4)
-  **Silhouette Score** calculation for cluster validation
-  **PCA** for linear dimensionality reduction and 2D visualization
-  **t-SNE** for nonlinear manifold visualization of cluster separation
-  **Stage distribution plots** for both clustering methods
-  Saves clustered dataset as `clustered_train_FD002.csv` and exports all visualizations to the `figures/` folder


In [2]:
# Clustering and Visualization for FD002
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score
import os

base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

df = pd.read_csv(os.path.join(base_path, "clean_train_FD002.csv"))

def stage_label(n):
    return f"Stage {n}"

print("\n⏳ Clustering and visualizing FD002...")
sensor_features = [col for col in df.columns if col.startswith("sensor_")]
X = df[sensor_features].copy()

kmeans = KMeans(n_clusters=5, random_state=42)
agglom = AgglomerativeClustering(n_clusters=5)
df['kmeans_cluster'] = kmeans.fit_predict(X)
df['agglo_cluster'] = agglom.fit_predict(X)
df['kmeans_stage'] = df['kmeans_cluster'].apply(stage_label)
df['agglo_stage'] = df['agglo_cluster'].apply(stage_label)

print("FD002 — KMeans Silhouette:", silhouette_score(X, df['kmeans_cluster']))
print("FD002 — Agglomerative Silhouette:", silhouette_score(X, df['agglo_cluster']))

# PCA
pca = PCA(n_components=2)
df[['pca_1', 'pca_2']] = pca.fit_transform(X)
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='kmeans_stage', palette='tab10')
plt.title("FD002: KMeans Stages via PCA")
plt.savefig(os.path.join(figures_path, "FD002_PCA_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='agglo_stage', palette='tab10')
plt.title("FD002: Agglomerative Stages via PCA")
plt.savefig(os.path.join(figures_path, "FD002_PCA_Agglo.png"))
plt.close()

# TSNE
tsne = TSNE(n_components=2, perplexity=30, max_iter=1000, random_state=42)
df[['tsne_1', 'tsne_2']] = tsne.fit_transform(X)
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='kmeans_stage', palette='tab10')
plt.title("FD002: KMeans Stages via t-SNE")
plt.savefig(os.path.join(figures_path, "FD002_TSNE_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='agglo_stage', palette='tab10')
plt.title("FD002: Agglomerative Stages via t-SNE")
plt.savefig(os.path.join(figures_path, "FD002_TSNE_Agglo.png"))
plt.close()

# Distribution
plt.figure(figsize=(10, 4))
sns.countplot(x='kmeans_stage', hue='kmeans_stage', data=df, palette='tab10')
plt.title("FD002 KMeans Stage Distribution")
plt.savefig(os.path.join(figures_path, "FD002_KMeans_Distribution.png"))
plt.close()

sns.countplot(x='agglo_stage', hue='agglo_stage', data=df, palette='tab10')
plt.title("FD002 Agglomerative Stage Distribution")
plt.savefig(os.path.join(figures_path, "FD002_Agglo_Distribution.png"))
plt.close()

df.to_csv(os.path.join(base_path, "clustered_train_FD002.csv"), index=False)
print("[✔] Saved clustered dataset: clustered_train_FD002.csv")



⏳ Clustering and visualizing FD002...
FD002 — KMeans Silhouette: 0.8840095392448586
FD002 — Agglomerative Silhouette: 0.8840095392448586
[✔] Saved clustered dataset: clustered_train_FD002.csv


# Phase 2: FD003 Clustering with KMeans and Agglomerative

This code performs clustering on the **FD003** dataset using both **KMeans** and **Agglomerative Clustering** with 5 clusters (mapped to Stages 0–4). The following steps are executed:

- Load and preprocess FD003 sensor data.
- Apply **KMeans** and **AgglomerativeClustering** on the sensor features.
- Compute **silhouette scores** to assess cluster separation quality.
- Reduce dimensions using **PCA** and **t-SNE** for 2D visualization.
- Generate plots:
  - PCA and t-SNE scatter plots for each clustering method.
  - Stage distribution bar plots.
- Save clustered data to: `clustered_train_FD003.csv`.
- All plots are saved under the `/figures` directory.

This forms the core of **Phase 2 clustering visualization** for FD003.


In [3]:
# Phase 2: FD003 Clustering with KMeans and Agglomerative
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score
import os

ds_id = 'FD003'
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

df = pd.read_csv(os.path.join(base_path, f"clean_train_{ds_id}.csv"))
sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
X = df[sensor_cols]

# Clustering
df['kmeans_cluster'] = KMeans(n_clusters=5, random_state=42).fit_predict(X)
df['agglo_cluster'] = AgglomerativeClustering(n_clusters=5).fit_predict(X)
df['kmeans_stage'] = df['kmeans_cluster'].apply(lambda n: f"Stage {n}")
df['agglo_stage'] = df['agglo_cluster'].apply(lambda n: f"Stage {n}")

# Silhouette
print(f"{ds_id} — KMeans Silhouette:", silhouette_score(X, df['kmeans_cluster']))
print(f"{ds_id} — Agglomerative Silhouette:", silhouette_score(X, df['agglo_cluster']))

# PCA
pca = PCA(n_components=2).fit_transform(X)
df['pca_1'], df['pca_2'] = pca[:, 0], pca[:, 1]
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='kmeans_stage')
plt.title(f"{ds_id}: KMeans via PCA")
plt.savefig(os.path.join(figures_path, f"{ds_id}_PCA_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='agglo_stage')
plt.title(f"{ds_id}: Agglomerative via PCA")
plt.savefig(os.path.join(figures_path, f"{ds_id}_PCA_Agglo.png"))
plt.close()

# t-SNE
tsne = TSNE(n_components=2, perplexity=30, random_state=42, n_iter=1000)
tsne_result = tsne.fit_transform(X)
df['tsne_1'], df['tsne_2'] = tsne_result[:, 0], tsne_result[:, 1]
sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='kmeans_stage')
plt.title(f"{ds_id}: KMeans via t-SNE")
plt.savefig(os.path.join(figures_path, f"{ds_id}_TSNE_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='agglo_stage')
plt.title(f"{ds_id}: Agglomerative via t-SNE")
plt.savefig(os.path.join(figures_path, f"{ds_id}_TSNE_Agglo.png"))
plt.close()

# Distribution
plt.figure(figsize=(10, 4))
sns.countplot(x='kmeans_stage', data=df, palette='tab10')
plt.title(f"{ds_id} KMeans Cluster Distribution")
plt.savefig(os.path.join(figures_path, f"{ds_id}_KMeans_Distribution.png"))
plt.close()

sns.countplot(x='agglo_stage', data=df, palette='tab10')
plt.title(f"{ds_id} Agglomerative Cluster Distribution")
plt.savefig(os.path.join(figures_path, f"{ds_id}_Agglo_Distribution.png"))
plt.close()

# Save
df.to_csv(os.path.join(base_path, f"clustered_train_{ds_id}.csv"), index=False)
print(f"[✔] Saved clustered dataset: clustered_train_{ds_id}.csv")


FD003 — KMeans Silhouette: 0.24536361705185084
FD003 — Agglomerative Silhouette: 0.22500565115374332


c:\Users\pandr\hybrid-predictive-maintenance\venv\Lib\site-packages\sklearn\manifold\_t_sne.py:1164: FutureWarning: 'n_iter' was renamed to 'max_iter' in version 1.5 and will be removed in 1.7.
  warnings.warn(
C:\Users\pandr\AppData\Local\Temp\ipykernel_8512\2921175764.py:61: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(x='kmeans_stage', data=df, palette='tab10')
C:\Users\pandr\AppData\Local\Temp\ipykernel_8512\2921175764.py:66: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  sns.countplot(x='agglo_stage', data=df, palette='tab10')


[✔] Saved clustered dataset: clustered_train_FD003.csv


## Phase 2: Clustering and Visualization for FD004

This block performs clustering on the FD004 dataset using **KMeans** and **Agglomerative Clustering**, followed by dimensionality reduction (PCA, t-SNE) for visual validation and stage distribution analysis.

### Key Steps:
- Load `clean_train_FD004.csv`
- Select only sensor features
- Apply KMeans and Agglomerative clustering (5 stages)
- Assign readable stage labels (`Stage 0` to `Stage 4`)
- Calculate silhouette scores to assess cluster quality
- Visualize clusters via:
  - **PCA** (Principal Component Analysis)
  - **t-SNE** (t-Distributed Stochastic Neighbor Embedding)
  - **Count distribution plots**
- Save the output as `clustered_train_FD004.csv` and all figures to `../figures/`

This clustering output will be used in later phases for classification and regression modeling.


In [4]:
# Clustering and Visualization for FD004
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE
from sklearn.cluster import KMeans, AgglomerativeClustering
from sklearn.metrics import silhouette_score
import os

base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

df = pd.read_csv(os.path.join(base_path, "clean_train_FD004.csv"))

def stage_label(n):
    return f"Stage {n}"

print("\n⏳ Clustering and visualizing FD004...")
sensor_features = [col for col in df.columns if col.startswith("sensor_")]
X = df[sensor_features].copy()

kmeans = KMeans(n_clusters=5, random_state=42)
agglom = AgglomerativeClustering(n_clusters=5)
df['kmeans_cluster'] = kmeans.fit_predict(X)
df['agglo_cluster'] = agglom.fit_predict(X)
df['kmeans_stage'] = df['kmeans_cluster'].apply(stage_label)
df['agglo_stage'] = df['agglo_cluster'].apply(stage_label)

print("FD004 — KMeans Silhouette:", silhouette_score(X, df['kmeans_cluster']))
print("FD004 — Agglomerative Silhouette:", silhouette_score(X, df['agglo_cluster']))

# PCA
pca = PCA(n_components=2)
df[['pca_1', 'pca_2']] = pca.fit_transform(X)
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='kmeans_stage', palette='tab10')
plt.title("FD004: KMeans Stages via PCA")
plt.savefig(os.path.join(figures_path, "FD004_PCA_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='pca_1', y='pca_2', hue='agglo_stage', palette='tab10')
plt.title("FD004: Agglomerative Stages via PCA")
plt.savefig(os.path.join(figures_path, "FD004_PCA_Agglo.png"))
plt.close()

# TSNE
tsne = TSNE(n_components=2, perplexity=30, max_iter=1000, random_state=42)
df[['tsne_1', 'tsne_2']] = tsne.fit_transform(X)
plt.figure(figsize=(12, 5))
sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='kmeans_stage', palette='tab10')
plt.title("FD004: KMeans Stages via t-SNE")
plt.savefig(os.path.join(figures_path, "FD004_TSNE_KMeans.png"))
plt.close()

sns.scatterplot(data=df, x='tsne_1', y='tsne_2', hue='agglo_stage', palette='tab10')
plt.title("FD004: Agglomerative Stages via t-SNE")
plt.savefig(os.path.join(figures_path, "FD004_TSNE_Agglo.png"))
plt.close()

# Distribution
plt.figure(figsize=(10, 4))
sns.countplot(x='kmeans_stage', hue='kmeans_stage', data=df, palette='tab10')
plt.title("FD004 KMeans Stage Distribution")
plt.savefig(os.path.join(figures_path, "FD004_KMeans_Distribution.png"))
plt.close()

sns.countplot(x='agglo_stage', hue='agglo_stage', data=df, palette='tab10')
plt.title("FD004 Agglomerative Stage Distribution")
plt.savefig(os.path.join(figures_path, "FD004_Agglo_Distribution.png"))
plt.close()

df.to_csv(os.path.join(base_path, "clustered_train_FD004.csv"), index=False)
print("[✔] Saved clustered dataset: clustered_train_FD004.csv")



⏳ Clustering and visualizing FD004...
FD004 — KMeans Silhouette: 0.8861873330096136
FD004 — Agglomerative Silhouette: 0.8861873330096136
[✔] Saved clustered dataset: clustered_train_FD004.csv


## Compute and Save Silhouette Scores (Phase 2 Summary)

This step evaluates the quality of clustering results using **Silhouette Scores** for both KMeans and AgglomerativeClustering across all FD00x datasets (FD001–FD004). It then saves the scores into a structured JSON file for future reference and reporting.

### What This Does
- Iterates through all clustered datasets (`clustered_train_FD00x.csv`).
- Checks for the presence of `kmeans_cluster` and `agglo_cluster` columns.
- Computes silhouette scores using `sklearn.metrics.silhouette_score`.
- Saves a summary of the scores to `report/silhouette_scores.json`.

### Why This Step Is Important
- Quantifies how well-separated the clusters are in high-dimensional sensor space.
- Helps verify that clustering in Phase 2 was meaningful.
- Provides metrics for selecting the better clustering method (KMeans vs Agglomerative) per dataset.
- Supports documentation and visualization in reports or dashboards.
- Avoids rerunning clustering evaluation during later phases.

### Output
- JSON File: `report/silhouette_scores.json`

JSON format:
```json
{
  "FD001": { "KMeans": 0.1880, "Agglomerative": 0.1760 },
  "FD002": { "KMeans": 0.8840, "Agglomerative": 0.8840 },
  "FD003": { "KMeans": 0.2454, "Agglomerative": 0.2250 },
  "FD004": { "KMeans": 0.8862, "Agglomerative": 0.8862 }
}



In [5]:
import pandas as pd
import os
import json
from sklearn.metrics import silhouette_score

# Define paths
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))
os.makedirs(report_path, exist_ok=True)

# Dataset IDs
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
silhouette_summary = {}

# Loop over datasets
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    if not os.path.exists(file_path):
        print(f"File missing: {file_path}")
        continue

    df = pd.read_csv(file_path)
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]

    if 'kmeans_cluster' not in df.columns or 'agglo_cluster' not in df.columns:
        print(f"Missing cluster labels in {ds_id}")
        continue

    X = df[sensor_cols]

    try:
        kmeans_score = silhouette_score(X, df['kmeans_cluster'])
        agglo_score = silhouette_score(X, df['agglo_cluster'])

        silhouette_summary[ds_id] = {
            'KMeans': round(kmeans_score, 4),
            'Agglomerative': round(agglo_score, 4)
        }

        print(f"{ds_id}: KMeans = {kmeans_score:.4f}, Agglomerative = {agglo_score:.4f}")

    except Exception as e:
        print(f"Error computing silhouette for {ds_id}: {e}")

# Save silhouette scores to JSON
output_path = os.path.join(report_path, "silhouette_scores.json")
with open(output_path, "w") as f:
    json.dump(silhouette_summary, f, indent=4)

print(f"Silhouette scores saved to: {output_path}")
print("Summary:", silhouette_summary)


FD001: KMeans = 0.1880, Agglomerative = 0.1760
FD002: KMeans = 0.8840, Agglomerative = 0.8840
FD003: KMeans = 0.2454, Agglomerative = 0.2250
FD004: KMeans = 0.8862, Agglomerative = 0.8862
Silhouette scores saved to: c:\Users\pandr\hybrid-predictive-maintenance\report\silhouette_scores.json
Summary: {'FD001': {'KMeans': np.float64(0.188), 'Agglomerative': np.float64(0.176)}, 'FD002': {'KMeans': np.float64(0.884), 'Agglomerative': np.float64(0.884)}, 'FD003': {'KMeans': np.float64(0.2454), 'Agglomerative': np.float64(0.225)}, 'FD004': {'KMeans': np.float64(0.8862), 'Agglomerative': np.float64(0.8862)}}


## Silhouette Score Summary Table (KMeans vs Agglomerative)

This table summarizes the silhouette scores computed during clustering for each dataset using both KMeans and AgglomerativeClustering. The silhouette score reflects how well-separated and cohesive the clusters are — a higher score indicates better-defined clusters.

| Dataset | KMeans Score | Agglomerative Score |
|---------|--------------|---------------------|
| FD001   | 0.1880       | 0.1760              |
| FD002   | 0.8840       | 0.8840              |
| FD003   | 0.2454       | 0.2250              |
| FD004   | 0.8862       | 0.8862              |

### Interpretation

- **FD002** and **FD004** show strong clustering performance, with silhouette scores above 0.88, indicating clear separation of degradation stages.
- **FD003** demonstrates moderate clustering quality, suggesting potential improvement through better feature engineering or preprocessing.
- **FD001** has the weakest clustering performance, possibly due to subtle or overlapping sensor patterns. This may require more advanced preprocessing, rolling statistics, or manual inspection in Phase 2.5.

These results guide model selection in future phases. Datasets with higher clustering scores are more suitable for supervised classification (Phase 3) and hybrid scoring (Phase 4).


---
## Generate Silhouette Score Summary Table in Markdown

This script reads precomputed silhouette scores from a JSON file (`silhouette_scores.json`) and formats them into a markdown table for easy readability and inclusion in reports.

### Purpose
- Automate the generation of a markdown summary table.
- Provide a consistent view of clustering performance across all datasets (FD001–FD004).
- Save the output in a markdown file (`silhouette_summary.md`) for documentation and reporting.

### Why This Step is Important
- Saves time by eliminating the need to manually format evaluation results.
- Ensures uniform documentation standards across the project.
- Helps compare clustering quality (KMeans vs Agglomerative) side-by-side.

### Output
- A table in markdown format stored at `report/silhouette_summary.md`, showing silhouette scores rounded to 4 decimal places.


In [6]:
import os
import json

# Define report path
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))
json_path = os.path.join(report_path, "silhouette_scores.json")
summary_md = os.path.join(report_path, "silhouette_summary.md")

# Check if JSON exists
if not os.path.exists(json_path):
    print("silhouette_scores.json not found. Please run the scoring script first.")
else:
    with open(json_path, "r") as f:
        silhouette_summary = json.load(f)

    with open(summary_md, "w", encoding="utf-8") as md:
        md.write("## Silhouette Score Summary (KMeans vs Agglomerative)\n\n")
        md.write("| Dataset | KMeans Score | Agglomerative Score |\n")
        md.write("|---------|--------------|---------------------|\n")
        for ds, scores in silhouette_summary.items():
            md.write(f"| {ds} | {scores['KMeans']:.4f} | {scores['Agglomerative']:.4f} |\n")

    print(f"Markdown summary saved to: {summary_md}")


Markdown summary saved to: c:\Users\pandr\hybrid-predictive-maintenance\report\silhouette_summary.md


---
## Identify Top 5 High-Variance Sensors (Per Dataset)

This step computes and stores the top 5 sensors (from each FD00x dataset) that exhibit the highest variance across the training data. Variance is a strong proxy for sensor informativeness, especially in unsupervised and classification settings.

### Purpose
- To detect which sensors contribute the most variation across engine cycles, which are likely to capture degradation patterns.
- Helps focus visualization and interpretation efforts (especially in Phase 2.5 and PCA/t-SNE plots).
- Lays the foundation for potential sensor selection or weighting during classification and regression modeling.

### Why This Step is Useful
- Reduces dimensionality concerns by spotlighting key sensors.
- Provides interpretability in sensor behavior across datasets.
- Assists in manual cluster verification and future visual analytics.

### Output
- A JSON file `top_variance_sensors.json` inside the `/report/` directory.
- Each key is a dataset (FD001–FD004), and its value is a dictionary of the top 5 sensor names and their variance values.


In [7]:
import os
import json
import pandas as pd

# Define dataset IDs and paths
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))
os.makedirs(report_path, exist_ok=True)

top_variance_summary = {}

for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    if not os.path.exists(file_path):
        print(f"File missing: {file_path}")
        continue

    df = pd.read_csv(file_path)
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    
    top_sensors = df[sensor_cols].var().sort_values(ascending=False).head(5).round(4)
    top_variance_summary[ds_id] = top_sensors.to_dict()

    print(f"\n{ds_id} - Top 5 High-Variance Sensors:")
    print(top_sensors)

# Save results to JSON
output_path = os.path.join(report_path, "top_variance_sensors.json")
with open(output_path, "w") as f:
    json.dump(top_variance_summary, f, indent=4)

print(f"\nTop variance sensor summary saved to: {output_path}")



FD001 - Top 5 High-Variance Sensors:
sensor_11    0.0253
sensor_12    0.0247
sensor_4     0.0231
sensor_2     0.0227
sensor_21    0.0223
dtype: float64

FD002 - Top 5 High-Variance Sensors:
sensor_16    0.2220
sensor_1     0.1283
sensor_19    0.1267
sensor_13    0.1246
sensor_12    0.1219
dtype: float64

FD003 - Top 5 High-Variance Sensors:
sensor_11    0.0294
sensor_12    0.0275
sensor_7     0.0271
sensor_17    0.0256
sensor_4     0.0232
dtype: float64

FD004 - Top 5 High-Variance Sensors:
sensor_16    0.2195
sensor_1     0.1288
sensor_19    0.1269
sensor_13    0.1248
sensor_2     0.1175
dtype: float64

Top variance sensor summary saved to: c:\Users\pandr\hybrid-predictive-maintenance\report\top_variance_sensors.json


---
## Stage-Time Profile Visualization (Phase 2)

This step plots the average time (engine cycle) spent in each degradation stage predicted by KMeans for all FD00x datasets.

### Purpose
- To understand how long engines typically remain in each degradation stage.
- Longer dwell time in early stages (like Stage 0 or 1) and shorter in critical stages (Stage 3 or 4) indicates realistic progression.

### Why This Step is Important
- Serves as a visual sanity check to confirm stage labels are ordered logically.
- Helps justify manual relabeling decisions in Phase 2.5 if stage durations appear reversed or inconsistent.
- Useful for industry stakeholders to estimate maintenance planning windows.

### Output
- Saved bar plots (`Stage_TimeProfile.png`) under the `figures` folder, one for each dataset.


In [19]:
import matplotlib.pyplot as plt
import pandas as pd
import os

# Define dataset identifiers
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']

# Define base path to data and figures
base_path = '../data'
figures_path = '../figures'
os.makedirs(figures_path, exist_ok=True)

# Plot average engine cycle per cluster stage
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    df = pd.read_csv(file_path)
    
    # Handle missing 'time' column
    if 'time' not in df.columns:
        df['time'] = df.index + 1

    # Compute average time per KMeans stage
    mean_time = df.groupby('kmeans_stage')['time'].mean().sort_index()

    # Plot
    plt.figure(figsize=(6, 4))
    mean_time.plot(kind='bar', color='skyblue')
    plt.title(f"{ds_id} – Avg Cycle by KMeans Stage")
    plt.ylabel("Average Time (Cycle)")
    plt.tight_layout()

    # Save figure and print path
    plot_filename = f"{ds_id}_Stage_TimeProfile.png"
    full_path = os.path.abspath(os.path.join(figures_path, plot_filename))
    plt.savefig(full_path)
    plt.close()
    print(f"{ds_id} stage-time profile saved at: {full_path}")


FD001 stage-time profile saved at: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD001_Stage_TimeProfile.png
FD002 stage-time profile saved at: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD002_Stage_TimeProfile.png
FD003 stage-time profile saved at: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD003_Stage_TimeProfile.png
FD004 stage-time profile saved at: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD004_Stage_TimeProfile.png


## Cluster Mean Sensor Summary Export (by KMeans Stage)

This step creates a CSV summary of the mean sensor readings (and time, if available) for each degradation stage (`kmeans_stage`) across all four FD00x datasets.

### Why This Is Important
- It helps explain what differentiates each degradation stage.
- Reveals which sensors show critical changes in later stages (e.g., temperature spikes, pressure drops).
- Used for reporting, model justification, and interpretability.

### Output
- One CSV per dataset saved to `report/cluster_summaries/`
- Format: `{FD00x}_ClusterSummary.csv`


In [11]:
import os
import pandas as pd

# Dataset IDs and Paths
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
summary_folder = os.path.abspath(os.path.join(os.getcwd(), '..', 'report', 'cluster_summaries'))
os.makedirs(summary_folder, exist_ok=True)

# Generate and save cluster-wise summaries
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    df = pd.read_csv(file_path)

    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    cols_to_use = sensor_cols + ['time'] if 'time' in df.columns else sensor_cols

    # Group by cluster stage and compute mean
    summary = df.groupby("kmeans_stage")[cols_to_use].mean().round(3)

    # Save summary to CSV
    output_file = os.path.join(summary_folder, f"{ds_id}_ClusterSummary.csv")
    summary.to_csv(output_file)

    print(f"Saved summary for {ds_id} to: {output_file}")


Saved summary for FD001 to: c:\Users\pandr\hybrid-predictive-maintenance\report\cluster_summaries\FD001_ClusterSummary.csv
Saved summary for FD002 to: c:\Users\pandr\hybrid-predictive-maintenance\report\cluster_summaries\FD002_ClusterSummary.csv
Saved summary for FD003 to: c:\Users\pandr\hybrid-predictive-maintenance\report\cluster_summaries\FD003_ClusterSummary.csv
Saved summary for FD004 to: c:\Users\pandr\hybrid-predictive-maintenance\report\cluster_summaries\FD004_ClusterSummary.csv


## Elbow Curves for All Datasets (KMeans)

This section plots the Within-Cluster Sum of Squares (WCSS) for different values of `k` (from 2 to 10) across all four datasets (FD001–FD004).

### Purpose

The elbow method helps identify a suitable number of clusters by measuring how tightly the data points group together as `k` increases. A noticeable drop in WCSS followed by a leveling-off suggests that adding more clusters beyond that point offers diminishing returns.

### Why This Matters

- Supports our decision to use **5 clusters** for KMeans across all datasets.
- Helps avoid overfitting (too many clusters) or underfitting (too few clusters).
- Adds credibility to our methodology by showing that the clustering configuration is backed by quantitative analysis.
- These plots can be referenced in both presentations and the final report to explain model selection decisions.

Each dataset’s elbow plot is saved in the `/figures/` folder and the corresponding WCSS values are summarized in `/report/elbow_summary.md` and `/report/elbow_scores.json`.

In [18]:
import os
import json
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.cluster import KMeans

# Dataset identifiers
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']

# Define local paths
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
report_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report'))

# Ensure output directories exist
os.makedirs(figures_path, exist_ok=True)
os.makedirs(report_path, exist_ok=True)

# Dictionary to store WCSS for each dataset
elbow_scores = {}

# Generate elbow curve for each dataset
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    df = pd.read_csv(file_path)
    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    X = df[sensor_cols]

    wcss = []
    for k in range(2, 11):
        model = KMeans(n_clusters=k, random_state=42, n_init='auto')
        model.fit(X)
        wcss.append(round(model.inertia_, 2))

    elbow_scores[ds_id] = wcss

    # Save elbow plot
    plt.figure(figsize=(6, 4))
    plt.plot(range(2, 11), wcss, marker='o', linestyle='-', color='teal')
    plt.title(f"Elbow Curve - {ds_id}")
    plt.xlabel("Number of Clusters (k)")
    plt.ylabel("WCSS (Within-Cluster Sum of Squares)")
    plt.grid(True)
    plt.tight_layout()
    plot_path = os.path.join(figures_path, f"{ds_id}_ElbowCurve.png")
    plt.savefig(plot_path)
    plt.close()
    print(f"{ds_id}: Elbow curve saved → {plot_path}")

# Save WCSS values to JSON
json_path = os.path.join(report_path, "elbow_scores.json")
with open(json_path, "w") as f:
    json.dump(elbow_scores, f, indent=4)
print(f"WCSS values saved to JSON → {json_path}")

# Save Markdown summary
md_path = os.path.join(report_path, "elbow_summary.md")
with open(md_path, "w") as f:
    f.write("# Elbow Curve Summary: WCSS for Each Dataset\n\n")
    f.write("This table lists the WCSS values for k=2 to 10.\n\n")
    f.write("| Dataset | " + " | ".join([f"k={k}" for k in range(2, 11)]) + " |\n")
    f.write("|---------|" + "--------|" * 9 + "\n")
    for ds_id in dataset_ids:
        row = f"| {ds_id} | " + " | ".join(map(str, elbow_scores[ds_id])) + " |\n"
        f.write(row)
    f.write("\n## Notes\n")
    f.write("- WCSS decreases with increasing clusters.\n")
    f.write("- The 'elbow point' helps choose the optimal k.\n")
    f.write("- For this project, k=5 is a good balance to reflect 5 degradation stages.\n")

print(f"Markdown summary saved → {md_path}")


FD001: Elbow curve saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD001_ElbowCurve.png
FD002: Elbow curve saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD002_ElbowCurve.png
FD003: Elbow curve saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD003_ElbowCurve.png
FD004: Elbow curve saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD004_ElbowCurve.png
WCSS values saved to JSON → c:\Users\pandr\hybrid-predictive-maintenance\report\elbow_scores.json
Markdown summary saved → c:\Users\pandr\hybrid-predictive-maintenance\report\elbow_summary.md


---
## Visualizing Hierarchical Structure via Dendrograms (Ward Linkage)

This step visualizes the hierarchical clustering structure of the CMAPSS datasets (FD001–FD004) using dendrograms derived from **Ward linkage**. Each dendrogram helps validate the grouping and structure of clusters formed in Phase 2.

### Purpose
- To examine the nested structure of the data and assess if 5 clusters is a reasonable choice.
- To support the unsupervised stage derivation with interpretable visual hierarchy.
- To reveal potential substructures or ambiguities in clustering.

### Methodology
- A random sample of 500 data points is selected from each dataset for computational feasibility.
- Ward’s method is used to compute the linkage matrix, optimizing for minimal intra-cluster variance.
- The dendrogram is truncated to the last 20 merges (`p=20`) for better clarity and readability.

### Output
- Dendrogram plots for each dataset saved in the `figures/` folder:
  - `FD001_Dendrogram.png`
  - `FD002_Dendrogram.png`
  - `FD003_Dendrogram.png`
  - `FD004_Dendrogram.png`

### Why This Step Is Important
- Confirms if KMeans and Agglomerative clusters reflect real nested structure.
- Helps justify the use of 5 clusters by locating the "merge height jumps" in dendrogram.
- Offers a secondary method of visual verification, strengthening the clustering phase’s robustness.



In [17]:
import os
import pandas as pd
import matplotlib.pyplot as plt
from scipy.cluster.hierarchy import linkage, dendrogram

# Dataset and path setup
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
figures_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
os.makedirs(figures_path, exist_ok=True)

# Loop through each dataset
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    df = pd.read_csv(file_path)

    sensor_cols = [col for col in df.columns if col.startswith("sensor_")]
    if len(df) < 500:
        raise ValueError(f"{ds_id} contains fewer than 500 samples. Consider using all data.")
    
    # Sample 500 rows for dendrogram clarity
    X_sample = df[sensor_cols].sample(n=500, random_state=42)

    # Compute linkage matrix using Ward’s method
    linkage_matrix = linkage(X_sample, method='ward')

    # Plot dendrogram
    plt.figure(figsize=(10, 4))
    dendrogram(linkage_matrix, truncate_mode='lastp', p=20)
    plt.title(f"Dendrogram - {ds_id} (Sample of 500)")
    plt.tight_layout()

    # Save the figure
    plot_path = os.path.join(figures_path, f"{ds_id}_Dendrogram.png")
    plt.savefig(plot_path)
    plt.close()

    print(f"Dendrogram saved → {plot_path}")


Dendrogram saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD001_Dendrogram.png
Dendrogram saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD002_Dendrogram.png
Dendrogram saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD003_Dendrogram.png
Dendrogram saved → c:\Users\pandr\hybrid-predictive-maintenance\figures\FD004_Dendrogram.png


---
## Top 5 High-Variance Sensor Trends by KMeans Stage

This section visualizes the average values of the top 5 high-variance sensors across each of the five KMeans stages (0 to 4) for all FD00x datasets. These sensors were selected based on their variance to help identify which parameters show the most degradation-specific variation.

### Purpose
- To understand how different sensors behave across degradation stages.
- To verify whether specific sensors show monotonic or pattern-based degradation over time.
- To guide feature selection for the classification model in Phase 3.

### Why Use Top 5?
- Top 5 sensors typically capture ~80–90% of variance seen across degradation patterns.
- Helps in narrowing down the dimensionality for explainability and faster model convergence.

### Output
Each plot shows 5 bars per sensor (for each degradation stage), saved under `figures/` as:

- `FD001_Top5Sensors_PerStage.png`
- `FD002_Top5Sensors_PerStage.png`
- `FD003_Top5Sensors_PerStage.png`
- `FD004_Top5Sensors_PerStage.png`


In [16]:
import os
import json
import pandas as pd
import matplotlib.pyplot as plt

# Dataset list and paths
dataset_ids = ['FD001', 'FD002', 'FD003', 'FD004']
base_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'data'))
fig_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'figures'))
json_path = os.path.abspath(os.path.join(os.getcwd(), '..', 'report', 'top_variance_sensors.json'))

# Ensure figure directory exists
os.makedirs(fig_path, exist_ok=True)

# Load top 5 variance sensors from JSON
with open(json_path, "r") as f:
    top_sensors = json.load(f)

# Generate and save plots
for ds_id in dataset_ids:
    file_path = os.path.join(base_path, f"clustered_train_{ds_id}.csv")
    if not os.path.exists(file_path):
        print(f"File not found: {file_path}")
        continue

    df = pd.read_csv(file_path)
    if "kmeans_stage" not in df.columns:
        print(f"Missing clustering column in {ds_id}")
        continue

    sensors_to_plot = list(top_sensors[ds_id].keys())[:5]
    mean_vals = df.groupby("kmeans_stage")[sensors_to_plot].mean()

    # Plotting
    mean_vals.plot(kind="bar", figsize=(10, 6))
    plt.title(f"{ds_id} – Top 5 Sensor Averages by Stage (KMeans)")
    plt.ylabel("Average Normalized Sensor Value")
    plt.xlabel("KMeans Stage")
    plt.xticks(rotation=0)
    plt.legend(title="Sensors")
    plt.tight_layout()

    # Save plot
    save_path = os.path.join(fig_path, f"{ds_id}_Top5Sensors_PerStage.png")
    plt.savefig(save_path)
    plt.close()
    print(f"Saved: {save_path}")


Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD001_Top5Sensors_PerStage.png
Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD002_Top5Sensors_PerStage.png
Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD003_Top5Sensors_PerStage.png
Saved: c:\Users\pandr\hybrid-predictive-maintenance\figures\FD004_Top5Sensors_PerStage.png
